### ISSUE TITLE
EC + x509 signing not working in 1.9?

### ISSUE THREAD

**Issue by chet** [pyca/cryptography#3773]
Howdy!

I have the following snippet:

```
from cryptography import x509
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives import (
    hashes,
    serialization,
)
from cryptography.hazmat.primitives.asymmetric import ec
from cryptography.x509.oid import NameOID

key = ec.generate_private_key(ec.SECP256R1(), default_backend())
csr = x509.CertificateSigningRequestBuilder().subject_name(x509.Name([
    x509.NameAttribute(NameOID.COUNTRY_NAME, 'US'),
    x509.NameAttribute(NameOID.STATE_OR_PROVINCE_NAME, 'CA'),
    x509.NameAttribute(NameOID.LOCALITY_NAME, 'San Francisco'),
    x509.NameAttribute(NameOID.ORGANIZATION_NAME, 'My Company'),
    x509.NameAttribute(NameOID.COMMON_NAME, 'mysite.com'),
])).sign(key, hashes.SHA256(), default_backend())
```

This has been working just fine, and today I updated to `cryptography==1.9`, and now it's failing with:

```
$ python example509.py 
Traceback (most recent call last):
  File "example509.py", line 18, in <module>
    ])).sign(key, hashes.SHA256(), default_backend())
  File "/Users/chet/virtualenvs/crypto-playground/lib/python2.7/site-packages/cryptography/x509/base.py", line 393, in sign
    return backend.create_x509_csr(self, private_key, algorithm)
  File "/Users/chet/virtualenvs/crypto-playground/lib/python2.7/site-packages/cryptography/hazmat/backends/openssl/backend.py", line 716, in create_x509_csr
    self._lib.RSA_R_DIGEST_TOO_BIG_FOR_RSA_KEY
  File "/Users/chet/virtualenvs/crypto-playground/lib/python2.7/site-packages/cryptography/hazmat/backends/openssl/backend.py", line 107, in openssl_assert
    return binding._openssl_assert(self._lib, ok)
  File "/Users/chet/virtualenvs/crypto-playground/lib/python2.7/site-packages/cryptography/hazmat/bindings/openssl/binding.py", line 75, in _openssl_assert
    errors_with_text
cryptography.exceptions.InternalError: Unknown OpenSSL error. This error is commonly encountered when another library is not cleaning up the OpenSSL error stack. If you are using cryptography with another library that uses OpenSSL try disabling it before reporting a bug. Otherwise please file an issue at https://github.com/pyca/cryptography/issues with information on how to reproduce this. ([])
```

I'm filing this issue, primarily because the exception message says to! That said, this might totally be on me; I tested it on a *different* machine with `cryptography==1.9`, and it works just fine.

The **non-working** machine has a *brew-installed* OpenSSL:

```
$ openssl version
OpenSSL 1.0.2l  25 May 2017
```

*(..however, after `pip` downgrading back to `cryptography==1.8`, the above snippet works again)*

The **working** machine has the *macOS Sierra-installed* OpenSSL:

```
$ openssl version
OpenSSL 0.9.8zh 14 Jan 2016
```

*(and that one works fine with both `cryptography==1.8` and `cryptography==1.9`)*

I don't want to send you on a wild goose chase if you think there's a good chance it's just my environment, but, I'm guessing there was something added in `cryptography==1.9` that does some better verification, and it's coming across something odd?

Also, if it helps:

```
$ python
Python 2.7.13 (default, Feb 28 2017, 13:40:05) 
[GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
```

```
$ pip list
asn1crypto (0.22.0)
cffi (1.10.0)
cryptography (1.9)
enum34 (1.1.6)
idna (2.5)
ipaddress (1.0.18)
pip (9.0.1)
pycparser (2.18)
setuptools (36.0.1)
six (1.10.0)
wheel (0.29.0)
```

Thank you in advance for any help and/or follow up!

**Comment by alex** [comment]
What version of `pip` did you use on each machine?

**Comment by chet** [comment]
Hey @alex! Both are using `pip` version 9.0.1.

**Comment by alex** [comment]
Can you run `python -c "from cryptography.hazmat.backends.openssl import backend; print(backend.openssl_version_text())"` in both environments?

**Comment by chet** [comment]
Sure!

Bad machine:
```
$ python -c "from cryptography.hazmat.backends.openssl import backend; print(backend.openssl_version_text())"
OpenSSL 1.0.2k  26 Jan 2017
```

Good machine:
```
$ python -c "from cryptography.hazmat.backends.openssl import backend; print(backend.openssl_version_text())"
OpenSSL 1.1.0f  25 May 2017
```

**Comment by alex** [comment]
That's interesting. And you're sure on the bad machine you got cryptography 1.9 from pip 9.0.1?

Can you show me the log of `pip install cryptography...`?

**Comment by chet** [comment]
Yeah :\

1) Making sure `pip==9.0.1`:

```
$ pip -V
pip 9.0.1 from /Users/chet/virtualenvs/crypto-playground/lib/python2.7/site-packages (python 2.7)
```

2) Checking currently installed packages:

```
$ pip list
pip (9.0.1)
setuptools (36.0.1)
wheel (0.29.0)
```

3) Installing `cryptography`:

```
$ pip install cryptography
Collecting cryptography
  Downloading cryptography-1.9.tar.gz (409kB)
    100% |████████████████████████████████| 419kB 2.5MB/s 
Collecting idna>=2.1 (from cryptography)
  Downloading idna-2.5-py2.py3-none-any.whl (55kB)
    100% |████████████████████████████████| 61kB 7.6MB/s 
Collecting asn1crypto>=0.21.0 (from cryptography)
  Downloading asn1crypto-0.22.0-py2.py3-none-any.whl (97kB)
    100% |████████████████████████████████| 102kB 8.1MB/s 
Collecting six>=1.4.1 (from cryptography)
  Downloading six-1.10.0-py2.py3-none-any.whl
Collecting enum34 (from cryptography)
  Downloading enum34-1.1.6-py2-none-any.whl
Collecting ipaddress (from cryptography)
  Downloading ipaddress-1.0.18-py2-none-any.whl
Collecting cffi>=1.7 (from cryptography)
  Downloading cffi-1.10.0-cp27-cp27mu-macosx_10_10_x86_64.whl (157kB)
    100% |████████████████████████████████| 163kB 5.6MB/s 
Collecting pycparser (from cffi>=1.7->cryptography)
  Downloading pycparser-2.18.tar.gz (245kB)
    100% |████████████████████████████████| 256kB 4.1MB/s 
Building wheels for collected packages: cryptography, pycparser
  Running setup.py bdist_wheel for cryptography ... done
  Stored in directory: /Users/chet/Library/Caches/pip/wheels/ff/a5/ef/186bb4f6a89ef0bb8373bf53e5c9884b96722f0857bd3111b8
  Running setup.py bdist_wheel for pycparser ... done
  Stored in directory: /Users/chet/Library/Caches/pip/wheels/95/14/9a/5e7b9024459d2a6600aaa64e0ba485325aff7a9ac7489db1b6
Successfully built cryptography pycparser
Installing collected packages: idna, asn1crypto, six, enum34, ipaddress, pycparser, cffi, cryptography
Successfully installed asn1crypto-0.22.0 cffi-1.10.0 cryptography-1.9 enum34-1.1.6 idna-2.5 ipaddress-1.0.18 pycparser-2.18 six-1.10.0
```

**Comment by chet** [comment]
So, I noticed I had a pending macOS update, so just I installed it, and now the *OpenSSL* backend version that `cryptography.hazmat.backends.openssl` sees has gone from `OpenSSL 1.0.2k  26 Jan 2017` back to `OpenSSL 0.9.8zh 14 Jan 2016`, haha! Still doesn't work, though.

Since 1.9 works just fine on the other machine, I'm going to say it's almost certainly something that I did with `openssl` and `homebrew`. I'm honestly not sure what, though, but I don't want to waste too much time on this if it has nothing to do with `cryptography`, and has more to do with some wonky environment stuff I have. :)

Without me digging through commits, do you know anything offhand that got added in `cryptography==1.9` re: verification that is probably the thing detecting I have something weird?

**Comment by alex** [comment]
Interesting. So for some reason your `pip` is downloading `tar.gz` instead of the wheel. The wheel would get you an OpenSSL 1.1.0.

This looks to be because we stopped publishing `cpy27mu` wheels for macOS.

How did you install your Python?

**Comment by chet** [comment]
Ahh..!

**Bad machine** (*has Python installed with homebrew*):
```
$ python
Python 2.7.13 (default, Feb 28 2017, 13:40:05) 
[GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)] on darwin
Type "help", "copyright", "credits" or "license" for more information.

$ ls -l $(which python)
lrwxr-xr-x  1 chet  admin  34 Feb 28 13:41 /usr/local/bin/python -> ../Cellar/python/2.7.13/bin/python
```



**Good machine** (*has Python installed with macOS*):
```
$ python
Python 2.7.10 (default, Feb  7 2017, 00:08:15) 
[GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.34)] on darwin
Type "help", "copyright", "credits" or "license" for more information

$ ls -l $(which python)
-rwxr-xr-x  1 root  wheel  66576 Apr 28 17:29 /usr/bin/python
```

Let me see if I can unwind some of this and confirm that's the issue. Pretty confident it will be!

**Comment by alex** [comment]
For the "Python installed with homebrew", can you check if there are any upgrades currently available?

cc: @tdsmith (trying to understand if it's expected that a homebrew python would need a `cp27mu` wheel?)

**Comment by chet** [comment]
So, I had done an `uninstall`/`unlink` on the *homebrew-installed* Python before you sent that, and confirmed that the system-level Python version worked fine. `pip` grabbed the `cryptography` wheel, as expected, and my code snippet worked as expected. :)

I then reinstalled the *homebrew* version, and when I installed `cryptography`, it downloaded the wheel (as expected) from:

https://pypi.python.org/packages/ca/6f/1f92293320881f98ba79bf976499e3f8089bcc62a3e967a6ef685af422c5/cryptography-1.9-cp27-cp27m-macosx_10_6_intel.whl

And it's working now. 🎉  I'm unsure if doing an upgrade would have yielded a new version, but I **did** notice that, while the *homebrew* Python is still `2.7.13`, it does have a newer build date than my previous *homebrew* build:

New:
```
$ python
Python 2.7.13 (default, Apr  4 2017, 08:47:57) 
[GCC 4.2.1 Compatible Apple LLVM 8.1.0 (clang-802.0.38)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
```

Previous:
```
$ python
Python 2.7.13 (default, Feb 28 2017, 13:40:05) 
[GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
```

And the OpenSSL backend is:

```
$ python -c "from cryptography.hazmat.backends.openssl import backend; print(backend.openssl_version_text())"
OpenSSL 1.1.0f  25 May 2017
```

I'm not sure if there was any good resolution here other than to reinstall Python. I would like to *think* that an update would have pulled in the new build?
